src/container/srv_target.c

```diff
@@ -2104,3 +2108,247 @@ ds_cont_oid_alloc_handler(crt_rpc_t *rpc)
 	out->co_map_version = 0;
 	crt_reply_send(rpc);
 }
+
+/* Track each container EC aggregation Epoch under ds_pool */
+struct cont_ec_eph {
+	uuid_t		ce_cont_uuid;
+	d_list_t	ce_list;
+	daos_epoch_t	ce_eph;
+	uint32_t	ce_updated:1;
+};
+
+/* Argument to query ec aggregate epoch from each xstream */
+struct cont_ec_xs_eph {
+	uuid_t		cont_uuid;
+	daos_epoch_t	eph;
+};
+
+struct cont_ec_xs_query_arg {
+	uuid_t			pool_uuid;
+	int			tgt_id;
+	int			ephs_cnt;
+	struct cont_ec_xs_eph	*ephs;
+};
+
+static int
+cont_ec_xs_reduce_alloc(struct dss_stream_arg_type *xs, void *agg_arg)
+{
+	struct cont_ec_xs_query_arg *xs_arg;
+	struct ds_pool *pool = agg_arg;
+
+	D_ALLOC_PTR(xs_arg);
+	if (xs_arg == NULL)
+		return -DER_NOMEM;
+
+	uuid_copy(xs_arg->pool_uuid, pool->sp_uuid);
+	xs->st_arg = xs_arg;
+	return 0;
+}
+
+static void
+cont_ec_xs_reduce_free(struct dss_stream_arg_type *xs)
+{
+	struct cont_ec_xs_query_arg *xs_arg = xs->st_arg;
+
+	if (xs_arg->ephs)
+		D_FREE_PTR(xs_arg->ephs);
+	D_FREE_PTR(xs_arg);
+}
+
+static struct cont_ec_eph *
+lookup_cont_ec_eph(d_list_t *ec_list, uuid_t cont_uuid)
+{
+	struct cont_ec_eph	*found = NULL;
+
+	d_list_for_each_entry(found, ec_list, ce_list) {
+		if (uuid_compare(found->ce_cont_uuid, cont_uuid) == 0)
+			return found;
+	}
+
+	return NULL;
+}
+
+static struct cont_ec_eph *
+lookup_insert_cont_ec_eph(d_list_t *ec_list, uuid_t cont_uuid)
+{
+	struct cont_ec_eph	*found;
+
+	found = lookup_cont_ec_eph(ec_list, cont_uuid);
+	if (found != NULL)
+		return found;
+	D_ALLOC_PTR(found);
+	if (found == NULL)
+		return NULL;
+
+	d_list_add(&found->ce_list, ec_list);
+	uuid_copy(found->ce_cont_uuid, cont_uuid);
+	return found;
+}
+
+static void
+cont_ec_eph_reduce(void *agg_arg, void *xs_arg)
+{
+	struct cont_ec_xs_query_arg	*x_arg = xs_arg;
+	struct ds_pool			*pool = agg_arg;
+	int				i;
+
+	for (i = 0; i < x_arg->ephs_cnt; i++) {
```

Comment:
This collective reduce call doesn't exclude the DOWN target, I think we need to exclude the DOWN targets, otherwise, the EC upper epoch for this rank could be always equal to the epoch of DOWN target? (Let's assume EC aggregation stop working on DOWN target)